src/main/scala/org/constellation/domain/redownload/RedownloadService.scala

```diff
@@ -1,38 +1,159 @@
+        case (id, recentSnaps) => recentSnaps.map(snap => (id, snap))
+      }
+    } yield deSerializedPeerProposals
+
+  private[redownload] def updateProposedSnapshots(fetchedProposals: Seq[(Id, RecentSnapshot)]) =
+    proposedSnapshots.modify { m =>
+      val updatedProposals = fetchedProposals.foldLeft(m) {
+        case (prevProposals, (id, recentSnap)) =>
+          val proposalsAtHeight = prevProposals.getOrElse(recentSnap.height, Map())
+          val hasMadeProposal = proposalsAtHeight.get(id).isDefined //todo Observation of new proposal
+          val test = prevProposals.updated(recentSnap.height, proposalsAtHeight + (id -> recentSnap))
+          if (!hasMadeProposal) test
+          else prevProposals
+      }
+      (updatedProposals, ())
+    }
 
-  // TODO: Check for the majority snapshot and store under a variable
-  def recalculateMajoritySnapshot(): F[Unit] = ???
+  def fetchAndSetPeerProposals() = fetchPeerProposals().flatMap(updateProposedSnapshots)
 
-  // TODO: Check for the alignment ownSnapshots <-> majoritySnapshot and call redownload if needed
-  def checkForAlignmentWithMajoritySnapshot(): F[Unit] = ???
+  def recalculateMajoritySnapshot(): F[(Seq[RecentSnapshot], Set[Id])] =
+    for {
+      allProposals <- proposedSnapshots.get
+      allProposalsNormalized = allProposals.toSeq.flatMap {
+        case (height, proposals: Map[Id, RecentSnapshot]) =>
+          proposals.map { case (id, recentSnapshot) => (id, Seq(recentSnapshot)) }.toSeq
+      }
+      peers <- cluster.readyPeers //todo need testing around join leave, this will cause cluster to stall if cluster.readyPeers not updated when peer leaves
+      (sortedSnaps, nodeIdsWithSnaps) = MajorityStateChooser.reDownloadPlan(allProposalsNormalized, peers.keySet.toList)
+    } yield (sortedSnaps, nodeIdsWithSnaps)
 
+  def checkForAlignmentWithMajoritySnapshot(): F[Option[List[RecentSnapshot]]] =
+    for {
+      peers <- cluster.readyPeers
+      majSnapsIds <- recalculateMajoritySnapshot()
+      ownSnaps <- getLocalSnapshots()
+      diff = MajorityStateChooser.compareSnapshotState(majSnapsIds, ownSnaps.toList)
+      shouldRedownload = RedownloadService.shouldReDownload(ownSnaps.toList, diff)
+      result <- if (shouldRedownload) {
+        logger.info(
+          s"[${cluster.id}] Re-download process with : \n" +
+            s"Snapshot to download : ${diff.snapshotsToDownload.map(a => (a.height, a.hash))} \n" +
+            s"Snapshot to delete : ${diff.snapshotsToDelete.map(a => (a.height, a.hash))} \n" +
+            s"From peers : ${diff.peers} \n" +
+            s"Own snapshots : ${ownSnaps.map(a => (a.height, a.hash))} \n" +
+            s"Major state : $majSnapsIds"
+        ) >>
+          healthChecker
+            .startReDownload(diff, peers.filterKeys(diff.peers.contains))
+            .flatMap(_ => Sync[F].delay[Option[List[RecentSnapshot]]](Some(majSnapsIds._1.toList)))
```

Comment:
- it's a pure value => why you need to make a flatMap and then delay?
- you don't need to specify a type here